### ISSUE TITLE
[HOLD for payment 2023-02-22] [$4000] Desktop - Chat - 4 lines are not appearing pasting with "cmd+v and cmd+shift+v

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#14243]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/13739

## Action Performed:
1. Go to ND and login
2. Click on any chat in the LHN
3. Copy text in the PR using "copy" icon
" I'm around if you are and wanna hop on a call https://meet.google.com/yjx-maas-pku

Nothing specific in mind"
4. Paste it without formatting into the chat composer user ctrl+shift+v and Paste it normally into the chat composer using ctrl+v
5. Copy the text by highlighting and Paste it without formatting into the chat composer user ctrl+shift+v and Paste it normally into the chat composer using ctrl+v

## Expected Result:
Verify that it still contains the 4 newlines and no more for both the methods

## Actual Result:
By step 4 it appears in 3 lines in the compose box and by step 5 appears with 3 lines

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Desktop

**Version Number:** 1.2.52.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/212099879-00e1e864-ff1a-4171-b0fa-0bb5572e2133.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ad5d26a1d71e0eec</li>
        <li>Upwork Job ID: 1614926800268017664</li>
        <li>Last Price Increase: 2023-01-30</li>
    </ul>
</details>

**Comment by tienifr** [comment]
# Proposal

This is due to we're replacing 2 lines with 1 line. To fix it, we need to remove that logic.

```diff
diff --git a/src/components/Composer/index.js b/src/components/Composer/index.js
index 5e00825c5a..ba0d2813e0 100755
--- a/src/components/Composer/index.js
+++ b/src/components/Composer/index.js
@@ -180,6 +180,7 @@ class Composer extends React.Component {
      * @param {String} text
      */
     paste(text) {
         try {
             document.execCommand('insertText', false, text);
             this.updateNumberOfLines();
@@ -210,6 +211,7 @@ class Composer extends React.Component {
     handlePaste(event) {
         event.preventDefault();
 
         const {files, types} = event.clipboardData;
         const TEXT_HTML = 'text/html';
 
@@ -219,6 +221,7 @@ class Composer extends React.Component {
             this.props.onPasteFile(event.clipboardData.files[0]);
             return;
         }
 
         // If paste contains HTML
         if (types.includes(TEXT_HTML)) {
@@ -268,7 +271,7 @@ class Composer extends React.Component {
             return;
         }
 
-        const plainText = event.clipboardData.getData('text/plain').replace(/\n\n/g, '\n');
+        const plainText = event.clipboardData.getData('text/plain');
 
         this.paste(Str.htmlDecode(plainText));
     }
```

**Comment by zanyrenney** [comment]
thanks @tienifr Will wait for review of your proposal, but in the meantime, please apply to the job on Upwork

**Comment by jatinsonijs** [comment]
@zanyrenney, @tienifr it is did intentionally to fix an issue. If we remove it. It will revert back following issue:
https://github.com/Expensify/App/issues/12769

**Comment by zanyrenney** [comment]
thanks for that additional context. @tgolen would love some further feedback here as to what the best path forward is? I'm going to put this issue and hold for now though.

**Comment by mallenexpensify** [comment]
@zanyrenney I'm removing the `Help Wanted` label if it's on hold, please add back when ready

**Comment by zanyrenney** [comment]
thanks @mallenexpensify - good shout!

**Comment by tgolen** [comment]
If you remove that regex, you're going to cause all the bugs to come back which that regex fixed. Please be sure to review the PR where I added that regex carefully. The fix for this bug needs to be done on top of the previous work.

**Comment by zanyrenney** [comment]
Thanks @tgolen - looks like we should be moving ahead with this one then with the caveat to review the PR from Tim carefully. Adding the help wanted label.

**Comment by mollfpr** [comment]
@zanyrenney Could you update the title too? Thanks!

**Comment by mollfpr** [comment]
@tylerkaraszewski @zanyrenney Is this still on hold or open for proposals?

**Comment by zanyrenney** [comment]
thanks for the bump @mollfpr - updated the title. If you're working on this one, can you please apply to the job on Upwork, I think this is you but not 100% sure from you GH handle / name. Thanks!

<img width="1067" alt="2023-01-24_15-56-53" src="https://user-images.githubusercontent.com/56830058/214342984-aef04fc5-19d8-4fc6-a6a6-5b45f6a198b3.png">

**Comment by mollfpr** [comment]
@zanyrenney Yup, that’s me!

**Comment by zanyrenney** [comment]
Brilliant, let me get you hired now!

**Comment by zanyrenney** [comment]
Have invited you @mollfpr - please accept the offer.

**Comment by mollfpr** [comment]
@zanyrenney Applied, thanks!

**Comment by mollfpr** [comment]
Still waiting for proposals...

**Comment by zanyrenney** [comment]
hey @mollfpr what about this proposal above? https://github.com/Expensify/App/issues/14243#issuecomment-1383812943

**Comment by mollfpr** [comment]
@zanyrenney The proposal is removing the regex that we need. Explain [here](https://github.com/Expensify/App/issues/14243#issuecomment-1385819163).

**Comment by TMisiukiewicz** [comment]
I'd like to investigate that issue, Callstack

**Comment by s77rt** [comment]
It looks safe to revert the change introduced by https://github.com/Expensify/App/pull/13739 (apply @tienifr proposal)
I have tested and I can't reproduce the initially bug reported here https://github.com/Expensify/App/issues/12769

**Comment by TMisiukiewicz** [comment]
found a potential issue - will do my best to post the proposal tomorrow.

**Comment by TMisiukiewicz** [comment]
## Proposal

### Issue

There is inconsistency between a way of copying text via context menu and `Cmd + C`. Messages copied via context menu are replacing single newlines with doubled and when pasting, it's replacing doubled with single ones. It does not seem to be necessary, as after parsing copied message to HTML the amount of newlines is correct. On the other hand, when copying highlighted text with `Cmd + C`, another function is called, which gets highlighted text using `SelectionScraper`. With this one, it contains additional spaces. Both of these scenarios worked because in the end, the paste function was replacing all the double newlines with single ones. However, when copying plain text from other source than app, it cannot go through `.replace(/\n\n/g, '\n')` and it can only access replacing in paste function. 

### Possible solution

Since copying via `Cmd + C` and context menu operate on different HTML output, and using `Cmd + C` is generating additional newlines and context menu not, we should replace all the `\n\n` with `\n` when using `Cmd + C`:

```diff
--- a/src/components/CopySelectionHelper.js
+++ b/src/components/CopySelectionHelper.js
@@ -36,7 +36,7 @@ class CopySelectionHelper extends React.Component {
             Clipboard.setString(parser.htmlToMarkdown(selection));
             return;
         }
-        Clipboard.setHtml(selection, Str.htmlDecode(parser.htmlToText(selection)));
+        Clipboard.setHtml(selection, Str.htmlDecode(parser.htmlToText(selection).replace(/\n\n/g, '\n')));
     }
```

And since the value of the copied message via context menu is correct from the beginning, it means it does not have to double newlines:

```diff
--- a/src/pages/home/report/ContextMenu/ContextMenuActions.js
+++ b/src/pages/home/report/ContextMenu/ContextMenuActions.js
@@ -112,11 +112,7 @@ export default [
                     if (!Clipboard.canSetHtml()) {
                         Clipboard.setString(parser.htmlToMarkdown(content));
                     } else {
-                        // Thanks to how browsers work, when text is highlighted and CTRL+c is pressed, browsers end up doubling the amount of newlines. Since the code in this file is
-                        // triggered from a context menu and not CTRL+c, the newlines need to be doubled so that the content that goes into the clipboard is consistent with CTRL+c behavior.
-                        // The extra newlines are stripped when the contents are pasted into the compose input, but if the contents are pasted outside of the comment composer, it will
-                        // contain extra newlines and that's OK because it is consistent with CTRL+c behavior.
-                        const plainText = Str.htmlDecode(parser.htmlToText(content)).replace(/\n/g, '\n\n');
+                        const plainText = Str.htmlDecode(parser.htmlToText(content));
                         Clipboard.setHtml(content, plainText);
                     }
                 }
```

Finally, since we do not have any doubled newlines when pasting, we don't have to replace anything when doing `Cmd + V`, or when copying selected text from the `Composer` component:

```diff
--- a/src/components/Composer/index.js
+++ b/src/components/Composer/index.js
@@ -268,7 +268,7 @@ class Composer extends React.Component {
             return;
         }
 
-        const plainText = event.clipboardData.getData('text/plain').replace(/\n\n/g, '\n');
+        const plainText = event.clipboardData.getData('text/plain');
 
         this.paste(Str.htmlDecode(plainText));
     }
@@ -297,9 +297,7 @@ class Composer extends React.Component {
         // the only stuff put into the clipboard is what the user selected.
         const selectedText = event.target.value.substring(this.state.selection.start, this.state.selection.end);
 
-        // The plaintext portion that is put into the clipboard needs to have the newlines duplicated. This is because
-        // the paste functionality is stripping all duplicate newlines to try and provide consistent behavior.
-        Clipboard.setHtml(selectedText, selectedText.replace(/\n/g, '\n\n'));
+        Clipboard.setHtml(selectedText, selectedText);
     }
```

Thanks to that, no matter the source of the copied text is, it is not modyfying any newlines except the ones when using copying highlighted text and `SelectionScraper`. Works correctly with pasting as markdown and plain text.

**Video using desktop app:**

https://user-images.githubusercontent.com/13985840/217511941-9f505d80-abe5-4f19-9589-d5b096956758.mov

**Video using web app** (the last two pastes come from cmd + C shortcut and it's not visible on the vid :) ):

https://user-images.githubusercontent.com/13985840/217511881-c59c7623-e70c-48a1-9a0d-713171a6d177.mov

**Comment by mountiny** [comment]
@TMisiukiewicz That is looking good! Have you tested on mobile apps too, please?

**Comment by TMisiukiewicz** [comment]
> @TMisiukiewicz That is looking good! Have you tested on mobile apps too, please?

@mountiny thanks for your comment - checked both mobile apps and web versions on mobile Chrome & Safari - everything works fine

**Comment by mollfpr** [comment]
Thanks, @TMisiukiewicz for the proposal and the clear explanation!

@mountiny The proposal looks good and tests well too 👍 Shall we go with the above proposal?

**Comment by mountiny** [comment]
I will defer to @tylerkaraszewski here to not take over from him!

**Comment by tgolen** [comment]
I see that this proposal covers two cases, but I believe there are at least seven cases total that we need to be sure to test. I put them all in the tests for this PR: https://github.com/Expensify/App/pull/13739. Can you please be sure all these are covered and work?

**Comment by TMisiukiewicz** [comment]
> I see that this proposal covers two cases, but I believe there are at least seven cases total that we need to be sure to test. I put them all in the tests for this PR: #13739. Can you please be sure all these are covered and work?

thanks for pointing that out - i will get back to you with the results once I test all the mentioned cases 👍

**Comment by mollfpr** [comment]
Here's the result from my end. Let me know if I miss the cases.

**Copy text normally**

https://user-images.githubusercontent.com/25520267/217576918-e92075df-f71b-4e6b-b744-b2a8d0c6538a.mov

**Copy text by right click and select copy to clipboard**

https://user-images.githubusercontent.com/25520267/217576968-07c6f701-8c83-4fa7-8780-4166bd21d48c.mov

**Copy text by hovered and select copy to clipboard**

https://user-images.githubusercontent.com/25520267/217577003-45f5fb4b-a89d-434f-9d97-50e0eac7d087.mov

**Copy text by highlighting the text message and right click then select copy to clipboard**

https://user-images.githubusercontent.com/25520267/217577043-8897ab30-1618-4f6e-b8b7-3dd36fdc64b8.mov


**Copy text inside the composer with ctrl+c**

https://user-images.githubusercontent.com/25520267/217577147-efadb925-3a0a-4676-8625-cc85dd832997.mov

**Copy text inside the composer with ctrl+x**


https://user-images.githubusercontent.com/25520267/217577177-1dd93059-a880-4092-8606-fad427e3f034.mov

**Comment by tgolen** [comment]
Thank you! That looks great 👍

**Comment by mountiny** [comment]
Agreed thanks for testing @mollfpr and nice job @TMisiukiewicz

**Comment by TMisiukiewicz** [comment]
@mountiny thank you! starting to work on a PR. Also thanks to @mollfpr for testing!

**Comment by MelvinBot** [comment]
@TMisiukiewicz @mollfpr @mountiny @zanyrenney this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @mollfpr is eligible for the Internal assigner, not assigning anyone new.

**Comment by mountiny** [comment]
All good here, PR is in review and should be hopefully resolved soon

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.71-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15034


If no regressions arise, payment will be issued on 2023-02-22. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mollfpr / @mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mollfpr / @mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mollfpr / @mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@TMisiukiewicz] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@zanyrenney] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mountiny** [comment]
@TMisiukiewicz Would you be able to compose the regression steps for this test as they were in the PR essentially but suited a bit more towards the QA team verify this bug hasnt come back?

@mollfpr would you be bale to help with checklist items? Thank you both!

**Comment by TMisiukiewicz** [comment]
**Regression test proposal**

- Go to URL https://staging.new.expensify.com/
- Log in with any account
- Navigate to any conversation
- Copy the following text using "Copy" button:
```
Lorem ipsum

dolor sit amet,


consectetur adipiscing
elit
```

- Paste it into Composer using `Cmd + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Paste it again using `Cmd + Shift + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Send the message on the conversation
- Highlight the sent message and copy it using `Cmd + C`
- Paste the message in the Composer using `Cmd + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Paste it again using `Cmd + Shift + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Highlight the sent message and copy it using right click and using "Copy to clipboard"
- Paste the message in the Composer using `Cmd + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Paste it again using `Cmd + Shift + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Right click on the last message and copy it using "Copy to clipboard"
- Paste the message in the Composer using `Cmd + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Paste it again using `Cmd + Shift + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Copy the sent message using "Copy" button from the context menu available when hovering on the message
- Paste the message in the Composer using `Cmd + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Remove the text from the composer
- Paste it again using `Cmd + Shift + V`
- Verify that the amount of newlines after pasting is the same as in the copied text 
- Do we agree 👍 or 👎

Well, that's a lot of steps but it covers all the cases 🤯  If needed, i can split it into separate scenarios to make it easier to test

**Comment by mollfpr** [comment]
> [@mollfpr / @mountiny] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/13739

> [@mollfpr / @mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/13739#issuecomment-1433196349

> [@mollfpr / @mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

https://expensify.slack.com/archives/C049HHMV9SM/p1676559147643779


Here's the checklist @mountiny

**Comment by zanyrenney** [comment]
@mountiny says hold fot payment for today. can I confirm I am good to go ahead with the payment? could you also clarify -- there is no timeframe bonus here, right?

**Comment by TomaszFrackowiak** [comment]
@zanyrenney - Hi we are from Callstack and you don't have to pay over Upwork, you just can close the issue.

**Comment by MelvinBot** [comment]
📣 @TomaszFrackowiak! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by zanyrenney** [comment]
thank you so much @TMisiukiewicz !

**Comment by mollfpr** [comment]
@zanyrenney You forget about me 😅

**Comment by mountiny** [comment]
@mollfpr asking internally, Zany is ooo for the rest of the week

**Comment by dylanexpensify** [comment]
@mollfpr I'm on it!

**Comment by mollfpr** [comment]
@dylanexpensify accepted ❤️

**Comment by dylanexpensify** [comment]
payment sent, contract ended my friend!
